internal/util/logging/mongo_handler.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"maps"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/FerretDB/FerretDB/v2/internal/util/must"
+)
+
+// mongoHandler is a [slog.Handler] that writes logs by using mongo structured JSON format.
+// The format returns log entries with Relaxed Extended JSON specification.
+// The format is not stable.
+//
+//nolint:vet // for readability
+type mongoHandler struct {
+	opts *NewHandlerOpts
+
+	goas      []groupOrAttrs
+	testAttrs map[string]any
+
+	m   *sync.Mutex
+	out io.Writer
+}
+
+// mongoLog represents a single log message in mongo structured JSON format.
+// For now some fields are ignored and may be empty.
+//
+//nolint:vet // to preserve field ordering
+type mongoLog struct {
+	Timestamp  primitive.DateTime `bson:"t"`
+	Severity   string             `bson:"s"`
+	Components string             `bson:"c"`
```

Comment:
I can see it's how `c`, `svc` and `id` are useful for parsing logs but we don't have useful value to populate. Our loggers are often initialized with `logging.WithName(l, "handler")` but that value is already available in `attr` :+1: